File: jmolecules-axon/02-axonstyle-cqrs-es-bank/src/main/java/org/jmolecules/example/axonframework/domain/api/exception/MaximumBalanceExceededException.java
package org.jmolecules.example.axonframework.domain.api.exception;
public class MaximumBalanceExceededException extends RuntimeException {


  public MaximumBalanceExceededException(String accountId, int currentBalance, int depositAmount, int maxAmount) {
    super(String.format("BankAccount[id=%s, currentBalance=%d]: Deposit of amount=%d not allowed, would exceed max. balance of %d",
      accountId,
      currentBalance,
      depositAmount,
      maxAmount
    ));
  }
}

File: jmolecules-axon/03-onion-axon-cqrs-es-bank/src/main/kotlin/org/jmolecules/example/axonframework/domain/api/command/MaximumBalanceExceededException.kt
package org.jmolecules.example.axonframework.domain.api.command

import org.jmolecules.example.axonframework.domain.api.type.AccountId
import org.jmolecules.example.axonframework.domain.api.type.Amount
import org.jmolecules.example.axonframework.domain.api.type.Balance

class MaximumBalanceExceededException(
    accountId: AccountId,
    currentBalance: Balance,
    maximumBalance: Balance,
    depositAmount: Amount?
) : RuntimeException(
    if (depositAmount == null) {
        "BankAccount[id=$accountId] can't be created, its balance $currentBalance would exceed max. balance of $maximumBalance"
    } else {
        "BankAccount[id=$accountId, currentBalance=$currentBalance]: Deposit of amount=$depositAmount not allowed, would exceed max. balance of $maximumBalance"
    }
)
